File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/HybridDownloadManager.kt
@file:Suppress("ktlint:standard:max-line-length")

package com.margelo.nitro.nitroplayer

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.NullType
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitroplayer.download.DownloadManagerCore

@DoNotStrip
@Keep
class HybridDownloadManager : HybridDownloadManagerSpec() {
    private val core: DownloadManagerCore

    init {
        val context =
            NitroModules.applicationContext
                ?: throw IllegalStateException("React Context is not initialized")
        core = DownloadManagerCore.getInstance(context)
    }

    // ── Configuration ─────────────────────────────────────────────────────────
    override fun configure(config: DownloadConfig) = core.configure(config)

    override fun getConfig(): DownloadConfig = core.getConfig()

    // ── Download operations ───────────────────────────────────────────────────
    override fun downloadTrack(
        track: TrackItem,
        playlistId: String?,
    ): Promise<String> = Promise.async { core.downloadTrack(track, playlistId) }

    override fun downloadPlaylist(
        playlistId: String,
        tracks: Array<TrackItem>,
    ): Promise<Array<String>> = Promise.async { core.downloadPlaylist(playlistId, tracks) }

    override fun pauseDownload(downloadId: String): Promise<Unit> = Promise.async { core.pauseDownload(downloadId) }

    override fun resumeDownload(downloadId: String): Promise<Unit> = Promise.async { core.resumeDownload(downloadId) }

    override fun cancelDownload(downloadId: String): Promise<Unit> = Promise.async { core.cancelDownload(downloadId) }

    override fun retryDownload(downloadId: String): Promise<Unit> = Promise.async { core.retryDownload(downloadId) }

    override fun pauseAllDownloads(): Promise<Unit> = Promise.async { core.pauseAllDownloads() }

    override fun resumeAllDownloads(): Promise<Unit> = Promise.async { core.resumeAllDownloads() }

    override fun cancelAllDownloads(): Promise<Unit> = Promise.async { core.cancelAllDownloads() }

    // ── Download status (sync) ────────────────────────────────────────────────
    override fun getDownloadTask(downloadId: String): Variant_NullType_DownloadTask {
        val task = core.getDownloadTask(downloadId)
        return if (task != null) {
            Variant_NullType_DownloadTask.create(task)
        } else {
            Variant_NullType_DownloadTask.create(NullType.NULL)
        }
    }

    override fun getActiveDownloads(): Array<DownloadTask> = core.getActiveDownloads()

    override fun getQueueStatus(): DownloadQueueStatus = core.getQueueStatus()

    override fun isDownloading(trackId: String): Boolean = core.isDownloading(trackId)

    override fun getDownloadState(trackId: String): DownloadState = core.getDownloadState(trackId)

    // ── Downloaded content queries (now async per spec) ───────────────────────
    override fun isTrackDownloaded(trackId: String): Promise<Boolean> = Promise.async { core.isTrackDownloaded(trackId) }

    override fun isPlaylistDownloaded(playlistId: String): Promise<Boolean> = Promise.async { core.isPlaylistDownloaded(playlistId) }

    override fun isPlaylistPartiallyDownloaded(playlistId: String): Promise<Boolean> = Promise.async { core.isPlaylistPartiallyDownloaded(playlistId) }

    override fun getDownloadedTrack(trackId: String): Promise<Variant_NullType_DownloadedTrack> =
        Promise.async {
            val track = core.getDownloadedTrack(trackId)
            if (track != null) {
                Variant_NullType_DownloadedTrack.create(track)
            } else {
                Variant_NullType_DownloadedTrack.create(NullType.NULL)
            }
        }

    override fun getAllDownloadedTracks(): Promise<Array<DownloadedTrack>> = Promise.async { core.getAllDownloadedTracks() }

    override fun getDownloadedPlaylist(playlistId: String): Promise<Variant_NullType_DownloadedPlaylist> =
        Promise.async {
            val playlist = core.getDownloadedPlaylist(playlistId)
            if (playlist != null) {
                Variant_NullType_DownloadedPlaylist.create(playlist)
            } else {
                Variant_NullType_DownloadedPlaylist.create(NullType.NULL)
            }
        }

    override fun getAllDownloadedPlaylists(): Promise<Array<DownloadedPlaylist>> = Promise.async { core.getAllDownloadedPlaylists() }

    override fun getLocalPath(trackId: String): Promise<Variant_NullType_String> =
        Promise.async {
            val path = core.getLocalPath(trackId)
            if (path != null) {
                Variant_NullType_String.create(path)
            } else {
                Variant_NullType_String.create(NullType.NULL)
            }
        }

    override fun syncDownloads(): Promise<Double> = Promise.async { core.syncDownloads().toDouble() }

    override fun getEffectiveUrl(track: TrackItem): Promise<String> = Promise.async { core.getEffectiveUrl(track) }

    // ── Deletion ──────────────────────────────────────────────────────────────
    override fun deleteDownloadedTrack(trackId: String): Promise<Unit> = Promise.async { core.deleteDownloadedTrack(trackId) }

    override fun deleteDownloadedPlaylist(playlistId: String): Promise<Unit> = Promise.async { core.deleteDownloadedPlaylist(playlistId) }

    override fun deleteAllDownloads(): Promise<Unit> = Promise.async { core.deleteAllDownloads() }

    override fun getStorageInfo(): Promise<DownloadStorageInfo> = Promise.async { core.getStorageInfo() }

    // ── Playback source ───────────────────────────────────────────────────────
    override fun setPlaybackSourcePreference(preference: PlaybackSource) = core.setPlaybackSourcePreference(preference)

    override fun getPlaybackSourcePreference(): PlaybackSource = core.getPlaybackSourcePreference()

    // ── Events ────────────────────────────────────────────────────────────────
    override fun onDownloadProgress(callback: (progress: DownloadProgress) -> Unit) = core.addProgressCallback(callback)

    override fun onDownloadStateChange(callback: (downloadId: String, trackId: String, state: DownloadState, error: DownloadError?) -> Unit) = core.addStateChangeCallback(callback)

    override fun onDownloadComplete(callback: (downloadedTrack: DownloadedTrack) -> Unit) = core.addCompleteCallback(callback)
}

File: react-native-nitro-player/ios/HybridDownloadManager.swift
//
//  HybridDownloadManager.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 2026-01-23..
//

import Foundation
import NitroModules

/// Hybrid implementation of DownloadManagerSpec for iOS
/// Bridges Nitro modules with the native DownloadManagerCore implementation
final class HybridDownloadManager: HybridDownloadManagerSpec {

  // MARK: - Properties

  private let core: DownloadManagerCore

  // MARK: - Initialization

  override init() {
    core = DownloadManagerCore.shared
    super.init()
  }

  // MARK: - Configuration

  func configure(config: DownloadConfig) throws {
    core.configure(config)
  }

  func getConfig() throws -> DownloadConfig {
    return core.getConfig()
  }

  // MARK: - Download Operations

  func downloadTrack(track: TrackItem, playlistId: String?) throws -> Promise<String> {
    return Promise.async {
      return self.core.downloadTrack(track: track, playlistId: playlistId)
    }
  }

  func downloadPlaylist(playlistId: String, tracks: [TrackItem]) throws -> Promise<[String]> {
    return Promise.async {
      return self.core.downloadPlaylist(playlistId: playlistId, tracks: tracks)
    }
  }

  // MARK: - Download Control

  func pauseDownload(downloadId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.pauseDownload(downloadId: downloadId)
    }
  }

  func resumeDownload(downloadId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.resumeDownload(downloadId: downloadId)
    }
  }

  func cancelDownload(downloadId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.cancelDownload(downloadId: downloadId)
    }
  }

  func retryDownload(downloadId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.retryDownload(downloadId: downloadId)
    }
  }

  func pauseAllDownloads() throws -> Promise<Void> {
    return Promise.async {
      self.core.pauseAllDownloads()
    }
  }

  func resumeAllDownloads() throws -> Promise<Void> {
    return Promise.async {
      self.core.resumeAllDownloads()
    }
  }

  func cancelAllDownloads() throws -> Promise<Void> {
    return Promise.async {
      self.core.cancelAllDownloads()
    }
  }

  // MARK: - Download Status

  func getDownloadTask(downloadId: String) throws -> Variant_NullType_DownloadTask {
    if let task = core.getDownloadTask(downloadId: downloadId) {
      return Variant_NullType_DownloadTask.second(task)
    }
    return Variant_NullType_DownloadTask.first(NullType.null)
  }

  func getActiveDownloads() throws -> [DownloadTask] {
    return core.getActiveDownloads()
  }

  func getQueueStatus() throws -> DownloadQueueStatus {
    return core.getQueueStatus()
  }

  func isDownloading(trackId: String) throws -> Bool {
    return core.isDownloading(trackId: trackId)
  }

  func getDownloadState(trackId: String) throws -> DownloadState {
    return core.getDownloadState(trackId: trackId) ?? .pending
  }

  // MARK: - Downloaded Content Queries

  func isTrackDownloaded(trackId: String) throws -> Promise<Bool> {
    Promise.async { self.core.isTrackDownloaded(trackId: trackId) }
  }

  func isPlaylistDownloaded(playlistId: String) throws -> Promise<Bool> {
    Promise.async { self.core.isPlaylistDownloaded(playlistId: playlistId) }
  }

  func isPlaylistPartiallyDownloaded(playlistId: String) throws -> Promise<Bool> {
    Promise.async { self.core.isPlaylistPartiallyDownloaded(playlistId: playlistId) }
  }

  func getDownloadedTrack(trackId: String) throws -> Promise<Variant_NullType_DownloadedTrack> {
    Promise.async {
      if let track = self.core.getDownloadedTrack(trackId: trackId) {
        return Variant_NullType_DownloadedTrack.second(track)
      }
      return Variant_NullType_DownloadedTrack.first(NullType.null)
    }
  }

  func getAllDownloadedTracks() throws -> Promise<[DownloadedTrack]> {
    Promise.async { self.core.getAllDownloadedTracks() }
  }

  func getDownloadedPlaylist(playlistId: String) throws -> Promise<Variant_NullType_DownloadedPlaylist> {
    Promise.async {
      if let playlist = self.core.getDownloadedPlaylist(playlistId: playlistId) {
        return Variant_NullType_DownloadedPlaylist.second(playlist)
      }
      return Variant_NullType_DownloadedPlaylist.first(NullType.null)
    }
  }

  func getAllDownloadedPlaylists() throws -> Promise<[DownloadedPlaylist]> {
    Promise.async { self.core.getAllDownloadedPlaylists() }
  }

  func getLocalPath(trackId: String) throws -> Promise<Variant_NullType_String> {
    Promise.async {
      if let path = self.core.getLocalPath(trackId: trackId) {
        return Variant_NullType_String.second(path)
      }
      return Variant_NullType_String.first(NullType.null)
    }
  }

  // MARK: - Deletion

  func deleteDownloadedTrack(trackId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.deleteDownloadedTrack(trackId: trackId)
    }
  }

  func deleteDownloadedPlaylist(playlistId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.deleteDownloadedPlaylist(playlistId: playlistId)
    }
  }

  func deleteAllDownloads() throws -> Promise<Void> {
    return Promise.async {
      self.core.deleteAllDownloads()
    }
  }

  // MARK: - Storage Management

  func getStorageInfo() throws -> Promise<DownloadStorageInfo> {
    return Promise.async {
      return self.core.getStorageInfo()
    }
  }

  func syncDownloads() throws -> Promise<Double> {
    Promise.async { Double(self.core.syncDownloads()) }
  }

  // MARK: - Playback Source Preference

  func setPlaybackSourcePreference(preference: PlaybackSource) throws {
    core.setPlaybackSourcePreference(preference)
  }

  func getPlaybackSourcePreference() throws -> PlaybackSource {
    return core.getPlaybackSourcePreference()
  }

  func getEffectiveUrl(track: TrackItem) throws -> Promise<String> {
    Promise.async { self.core.getEffectiveUrl(track: track) }
  }

  // MARK: - Event Callbacks

  func onDownloadProgress(callback: @escaping (DownloadProgress) -> Void) throws {
    NitroPlayerLogger.log("HybridDownloadManager", "onDownloadProgress callback registered")
    core.addProgressCallback(callback)
  }

  func onDownloadStateChange(
    callback: @escaping (String, String, DownloadState, DownloadError?) -> Void
  ) throws {
    NitroPlayerLogger.log("HybridDownloadManager", "onDownloadStateChange callback registered")
    core.addStateChangeCallback(callback)
  }

  func onDownloadComplete(callback: @escaping (DownloadedTrack) -> Void) throws {
    NitroPlayerLogger.log("HybridDownloadManager", "onDownloadComplete callback registered")
    core.addCompleteCallback(callback)
  }
}
